src/units.rs

```diff
@@ -1,116 +1,109 @@
+    round: u16,
+    creator: NodeIndex,
 }
 
-impl<H: Hasher, D: Data> FullUnit<H, D> {
-    pub(crate) fn creator(&self) -> NodeIndex {
-        self.inner.creator
-    }
-    pub(crate) fn round(&self) -> Round {
-        self.inner.round()
+impl UnitCoord {
+    pub fn new(round: Round, creator: NodeIndex) -> Self {
+        Self {
+            creator,
+            round: round as u16,
+        }
     }
 
-    pub(crate) fn coord(&self) -> UnitCoord {
-        (self.round(), self.creator()).into()
+    pub fn creator(&self) -> NodeIndex {
+        self.creator
     }
 
-    pub(crate) fn hash(&self) -> H::Hash {
-        H::hash(&self.encode())
+    pub fn round(&self) -> Round {
+        self.round as Round
     }
 }
 
-impl<H: Hasher, D: Data> Signable for FullUnit<H, D> {
-    type Hash = H::Hash;
-    fn hash(&self) -> H::Hash {
-        self.using_encoded(H::hash)
-    }
+/// Combined hashes of the parents of a unit together with the set of indices of creators of the
+/// parents
+#[derive(Clone, Debug, Default, PartialEq)]
+pub(crate) struct ControlHash<H: Hasher> {
+    pub(crate) parents_mask: bit_vec::BitVec<u32>,
+    pub(crate) hash: H::Hash,
 }
 
-impl<H: Hasher, D: Data> Index for FullUnit<H, D> {
-    fn index(&self) -> NodeIndex {
-        self.inner.creator
+impl<H: Hasher> ControlHash<H> {
+    pub(crate) fn new(parent_map: &NodeMap<Option<H::Hash>>) -> Self {
+        let hash = Self::combine_hashes(&parent_map);
+        let parents = parent_map.iter().map(Option::is_some).collect();
+
+        ControlHash {
+            parents_mask: parents,
+            hash,
+        }
     }
-}
 
-pub(crate) type UncheckedSignedUnit<H, D, S> = UncheckedSigned<FullUnit<H, D>, S>;
+    pub(crate) fn combine_hashes(parent_map: &NodeMap<Option<H::Hash>>) -> H::Hash {
+        parent_map.using_encoded(H::hash)
+    }
 
-pub(crate) type SignedUnit<'a, H, D, KB> = Signed<'a, FullUnit<H, D>, KB>;
+    pub(crate) fn parents(&self) -> impl Iterator<Item = NodeIndex> + '_ {
+        self.parents_mask
+            .iter()
+            .enumerate()
+            .filter(|(_, b)| *b)
+            .map(|(i, _)| i.into())
+    }
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq, Encode, Decode, StdHash)]
-pub(crate) struct UnitCoord {
-    pub(crate) creator: NodeIndex,
-    pub(crate) round: u64,
-}
+    pub(crate) fn n_parents(&self) -> NodeCount {
+        NodeCount(self.parents().count())
+    }
 
-impl<H: Hasher> From<Unit<H>> for UnitCoord {
-    fn from(unit: Unit<H>) -> Self {
-        UnitCoord {
-            creator: unit.creator(),
-            round: unit.round() as u64,
-        }
+    pub(crate) fn n_members(&self) -> NodeCount {
+        NodeCount(self.parents_mask.len())
     }
 }
 
-impl<H: Hasher> From<&Unit<H>> for UnitCoord {
-    fn from(unit: &Unit<H>) -> Self {
-        UnitCoord {
-            creator: unit.creator(),
-            round: unit.round() as u64,
-        }
+impl<H: Hasher> Encode for ControlHash<H> {
```

Comment:
Add a test checking if composition of encode and decode is an identity.

Reply:
Done

Reply:
I misunderstood you intially, and added such a test for FullUnit. Now there is a test for ControlHash as well.